### PR TITLE
Import History

### PR DESCRIPTION
**Pull request by blackforestboi**
i made some changes, now at least we have the url_html parsed into a real DOM tree. this should help.

### CODE REVIEW

**Review by bohrium272 — CHANGES_REQUESTED** [review]
Please change the loop statement at line 372 to `(0, history.length)`

**Comment by blackforestboi** [comment]
oh yes! had that for experimental purposes. Thanks for correcting. Fixed!

File: extension/js/download_History.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+        // check if user cancelled download
+        xhttp.onreadystatechange = function() {
+            if (abort_status == 1) {
+                localStorage['list_downloaded_urls'] = JSON.stringify(existing_urls);
+                //localStorage.setItem('abort_status',0)
+                document.getElementById("title_download").innerHTML = "Download cancelled!";
+                document.getElementById("close_message").innerHTML = "You can always restart it via the settings.";
+                document.getElementById("info_text").innerHTML = '<a href="preferences.html">Go to settings</a>';
+                document.getElementById('abort_button').remove()
```
Comment (obsidianart) [inline extension/js/download_History.js]: to do it properly you need to remove the click event first or you will create memory leak

File: extension/js/background.js [hunk 2]
```diff
@@ -123,6 +123,7 @@ function handleMessage(data, sender, sendRespones) {
     // data is from message
     if (data.msg === 'pageContent' && shouldArchive(data)) {
         delete data.msg;
+        console.log(data.text)
```
Comment (obsidianart) [inline extension/js/background.js]: needed?

File: extension/js/download_History.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
+var initial = document.body.parentNode.innerHTML;
+
+//open list of already indexed urls from local storage
+var existing_urls = JSON.parse(localStorage['list_downloaded_urls']);
+var abort_status = 0
```
Comment (obsidianart) [inline extension/js/download_History.js]: if is only 0 or 1 better to say isAbortedByUser = false

File: extension/assets/download_history.html [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        </div>
+
+  <h2 id="title_download">Step 2: Downloading History</h2>
+
+  <h3 id="close_message">Hang in there!</h3>
+
+  <p id="info_text">To make all your previously visited pages available for full-text search, we need to download them first. <br>Depending on the number of pages, this might take a while and might influence your browsing speed.<br> <strong>To cancel the download, deactivate and activate the WorldBrain extension.</strong></p>
+
+
+<button id="abort_button" style="float:right" class="ui red button">Cancel Download</button>
```
Comment (obsidianart) [inline extension/assets/download_history.html]: I would say this stop downloading, doesn't cancel it as you are saving what you downloading till that point

File: extension/js/background.js [hunk 5]
```diff
@@ -135,15 +136,15 @@ function handleMessage(data, sender, sendRespones) {
         timeIndex.push(time.toString());
```
Comment (obsidianart) [inline extension/js/background.js:135]: I don't see the point of changing time toString(), numbers are as good (maybe there's a reason I can't understand)

File: extension/js/download_History.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        // check if user cancelled download
+        xhttp.onreadystatechange = function() {
+            if (abort_status == 1) {
+                localStorage['list_downloaded_urls'] = JSON.stringify(existing_urls);
+                //localStorage.setItem('abort_status',0)
+                document.getElementById("title_download").innerHTML = "Download cancelled!";
+                document.getElementById("close_message").innerHTML = "You can always restart it via the settings.";
+                document.getElementById("info_text").innerHTML = '<a href="preferences.html">Go to settings</a>';
+                document.getElementById('abort_button').remove()
+                console.log("Cancelled Download")
```
Comment (obsidianart) [inline extension/js/download_History.js]: "Downloaded cancelled" (but I would say is stopped not cancelled)

File: extension/js/download_History.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                        //requesting HTML code of URL and converting it into DOM tree
+                        url_html = xhttp.responseText;
+                        var doc = document.implementation.createHTMLDocument();
+                        doc.documentElement.innerHTML = url_html
+                        
+                        // getting only the content with a div tag (therefore exluding script AND META (unfortunately) tags etc.)
+                        // for this whole thing we need a better solution because many divs actually have the same textContent and 
+                        //therefore we have things multiple times -> unecessary bloating of file, now ca 150kb, could be 30kb. 
+                        body_text = doc.body
+                        divs = body_text.getElementsByTagName('div')
```
Comment (obsidianart) [inline extension/js/download_History.js]: are you assuming the content is inside div? what about article, section, main and so on?

Comment (obsidianart) [reply extension/js/download_History.js]: We should go in this direction or find an open version of this https://www.readability.com/developers/api/parser

**Review by bohrium272 — APPROVED** [review]

File: extension/js/download_History.js (hunk 3 above)
Comment (blackforestboi) [reply extension/js/download_History.js]: done.

File: extension/js/download_History.js (hunk 1 above)
Comment (blackforestboi) [reply extension/js/download_History.js]: I don't understand that one.

File: extension/js/download_History.js (hunk 6 above)
Comment (blackforestboi) [reply extension/js/download_History.js]: Done

File: extension/js/download_History.js (hunk 7 above)
Comment (blackforestboi) [reply extension/js/download_History.js]: Actually we use readablity already for parsing the article before saving it. Will test out if we can use the local version we have. 

@arpitgogia said it only return parts of the text but that was before the adjustments to the correct DOM tree of the HTML file.

File: extension/js/background.js (hunk 2 above)
Comment (blackforestboi) [reply extension/js/background.js]: removed

File: extension/js/download_History.js (hunk 1 above)
Comment (obsidianart) [reply extension/js/download_History.js]: at line 8 you have

```
document.getElementById('abort_button').onclick = function(){
    abort_status = 1
};
```

which look like a bind but I just realized is not (never saw this tecnique before), you can just ignore this comment

File: extension/assets/download_history.html (hunk 4 above)
Comment (blackforestboi) [reply extension/assets/download_history.html]: legacy piece. Forgot to remove it. was before I added the cancel button. Thanks!

File: extension/js/download_History.js (hunk 7 above)
Comment (blackforestboi) [reply extension/js/download_History.js]: readabilities library is also not ideal. The browser itself with document.body.textContent is the best way to retrieve the content.

I have not found a way to extract just the visible textual content from an article with a JS library.

**Comment by blackforestboi** [comment]
closing as being obsolete with #27

**Pull request closed without merging**